nestkernel/sparse_node_array.h

```diff
@@ -39,52 +39,100 @@ namespace nest
 class Node;
 
 /**
- * Provide sparse representation of local nodes.
+ * Sparse representation of local nodes.
  *
- * This class is a container providing lookup of local nodes (as Node*)
- * based on node IDs.
+ * SparseNodeArray maps node IDs of thread-local nodes to Node*. It provides
+ * a const iterator interface for iteration over all local nodes, lookup by
+ * node ID and lookup by numeric index into local nodes. The latter is provided
+ * to support HPC synapses using TargetIdentifierIndex representation.
  *
- * Basically, this array is a vector containing only pointers to local nodes.
- * For M MPI processes, we have
+ * For efficient lookup, all normal nodes (with proxies) and all devices need
+ * to be created en bloc. It does not matter which are created first. If nodes
+ * and devices are created alternatingly, lookup performance may suffer significantly
+ * if many devices are present. This mainly affects network connection.
  *
- *   node ID  %  M  --> rank
- *   node ID div M  --> index on rank
+ * As nodes are added to an initially empty SparseNodeArray, the array tracks
+ * whether nodes have proxies or not. Array lookup is split at the point of the first
+ * proxy/no-proxy (or no-proxy/proxy) transition between nodes. In the no-proxy
+ * region, where all nodes are represented locally, Node IDs are mapped directly
+ * to array indices. In the proxy region, they are scaled by 1/n_vp.
+ *
+ * To reliably reject requests for node IDs beyond the globally maximal node ID, the
+ * latter must be set explicit. A SparseNodeArray is said to be in *consistent state*
+ * if the global maximal node ID has been set. Once add_local_node() is called, the
+ * array is not in consistent state until the global maximal node ID is set again. This
+ * is indicated by setting the max_node_id_ == 0. Looking up nodes while the
+ * array is not in consistent state triggers an assertion.
+ *
+ * To also support cases in which users interleave creation of nodes with and
+ * without proxies or use nodes with special behavior (music nodes, e.g.), we
+ * perform a linear search from the estimated location of the node in the array.
+ *
+ * The following invariants hold when the array is in consistent state:
```

Comment:
Is the term 'invariant' fairly well understood and used among developers? Would something like "the following points hold true / remain unchanged" be more clear?

Reply:
Yes, invariants are standard terminology in algorithm analysis.

Reply:
ok thanks